OBSDOCS-1359_New and OBSDOCS-817 (partially covered) (#83850)
openshift/openshift-docs: observability/logging/logging-6.0/log6x-about.adoc

Author: Thanks for sharing your thoughts. I incorporated first paragraph as per your suggestion. Yes, second paragraph suggestion seems to be out of scope for this PR.

== Validation
Logging includes extensive validation rules and default values to ensure a smooth and error-free configuration experience. The `ClusterLogForwarder` resource enforces validation checks on required fields, dependencies between fields, and the format of input values. Default values are provided for certain fields, reducing the need for explicit configuration in common scenarios.

== Quick Start

OpenShift Logging supports two data models:

* ViaQ (General Availability)
* OTEL (Technology Preview)

> You can select either of these data models based on your requirements. The ViaQ data model is the default and fully supported option, while OpenTelemetry (OTEL) uses it's standards and is not fully supported. For more information, see xref:../../../observability/logging/logging-6.0/log6x-otel-data-model.adoc#log6x-otel-data-model[OTEL data model].

Reviewer: I suggest to add a note here that in future releases of OpenShift Logging we will switch the default from ViaQ to OTEL.

Author: Updating as https://github.com/openshift/openshift-docs/pull/83850#discussion_r1811118183

> === Quick Start with ViaQ

Reviewer: I suggest to add a note at the end of this section after the `ClusterLogForwarder` resource explaining that the field `lokistack.dataModel: Viaq` is the current default.

== Validation
Logging includes extensive validation rules and default values to ensure a smooth and error-free configuration experience. The `ClusterLogForwarder` resource enforces validation checks on required fields, dependencies between fields, and the format of input values. Default values are provided for certain fields, reducing the need for explicit configuration in common scenarios.

> == Quick Start

Author: Added unique IDs for in scope changes for this PR.

OpenShift Logging supports two data models:

* ViaQ (General Availability)
* OpenTelemetry (Technology Preview)

You can select either of these data models based on your requirement by configuring the `lokiStack.dataModel` field in the `ClusterLogForwarder`. ViaQ is the default data model when forwarding logs to LokiStack. For more information on OpenTelemetry, see xref:../../../observability/logging/logging-6.0/log6x-opentelemetry-data-model.adoc#log6x-opentelemetry-data-model[OpenTelemetry data model].

[NOTE]
====
In future releases of OpenShift Logging, the default data model will change from ViaQ to OpenTelemetry.
====

=== Quick Start with ViaQ

To use the default ViaQ data model, follow these steps:  

.Prerequisites
* Cluster administrator permissions

.Procedure

> . Install the `Red Hat Openshift Logging Operator`, `Loki Operator`, and `Cluster Observability Operator` operators from OperatorHub.

Author: Incorporated where I can. Thanks!!

== Validation
Logging includes extensive validation rules and default values to ensure a smooth and error-free configuration experience. The `ClusterLogForwarder` resource enforces validation checks on required fields, dependencies between fields, and the format of input values. Default values are provided for certain fields, reducing the need for explicit configuration in common scenarios.

== Quick Start

OpenShift Logging supports two data models:

* ViaQ (General Availability)

> * OTEL (Technology Preview)

Reviewer 2: Should we call it "OpenTelemetry" here?

> You can select either of these data models based on your requirements. The ViaQ data model is the default and fully supported option, while OTEL uses the OpenTelemetry standards and is not fully supported.

Author: Please see: https://github.com/openshift/openshift-docs/pull/83850#discussion_r1812437303

=== Quick Start with ViaQ

To use the default ViaQ data model, follow these steps:  

[NOTE]
====
The ViaQ data model is now deprecated and is planned to be removed in future logging releases.

> ====

Reviewer 3: I'm afraid that's not right. IIRC ViaQ will only be deprecated once we GA OTEL

Reviewer 4: +1   Additionally, a few lines above this is the statement "The ViaQ data model is the default and fully supported option..."

Author: Considering above suggestion for the Note, I believe it would be best to remove the Note for now to avoid confusion. Once OTEL reaches GA, then we can include a formal deprecation notice. Let me know if you have other thoughts. Thanks!

$ oc create sa collector -n openshift-logging
----

> . Bind the `ClusterRole` to the service account:

Reviewer 2: Is it better to instead describe what the role is for?

```suggestion
. Allow the collector's service account to write data to the LokiStack:
```

> . Allow the collector's service account to write data to the LokiStack:

Author: Updated as, "Allow the collector's service account to write data to the `LokiStack` CR:"

+
[source,shell]
----
$ oc adm policy add-cluster-role-to-user logging-collector-logs-writer -z collector
----
+
[NOTE]
====
The `ClusterRole` is created automatically during the Cluster Logging Operator installation and does not need to be created manually.
====

> . Add additional roles to the collector service account:

Reviewer 2: Is it better to instead describe what the role is for?

```suggestion
. Allow the collector's service account to collect logs:
```

$ oc create sa collector -n openshift-logging
----

> . (Optional) Create a `ClusterRole` for the collector:

Author: Alright! I will delete this step then.

      name: logging-loki
----

. Create a `ClusterLogForwarder` CR to configure log forwarding:
+
[source,yaml]
----
apiVersion: observability.openshift.io/v1
kind: ClusterLogForwarder
metadata:
  name: collector
  namespace: openshift-logging
spec:
  serviceAccount:
    name: collector
  outputs:
    - name: default-lokistack
      type: lokiStack
      lokiStack:
        target:
          name: logging-loki
          namespace: openshift-logging

>         dataModel: ViaQ # <1>

Reviewer 2: I'm not sure we should specify `dataModel: ViaQ` in the example, because it will be copied as-is without thinking about it and then keep customers stuck in the old model, even when we eventually deprecate it.

If this would be on a special "try out the new features" page then it would be a different matter, but this is still the first page everyone looking at 6.1 will read, so the "default example" should be as minimal as possible.

>     - name: default-lokistack
>       type: lokiStack
>       lokiStack:
>         target:
>           name: logging-loki
>           namespace: openshift-logging
>         dataModel: ViaQ # <1>
>         authentication:
>           token:
>             from: serviceAccount
>       tls:
>         ca:
>           key: service-ca.crt
>           configMapName: openshift-service-ca.crt

Reviewer 2: I think you misinterpreted my previous comment as to change the indentation of the blocks. That's not what I meant by it, I'm sorry.

```suggestion
  - name: default-lokistack
    type: lokiStack
    lokiStack:
      target:
        name: logging-loki
        namespace: openshift-logging
      dataModel: ViaQ # <1>
      authentication:
        token:
          from: serviceAccount
    tls:
      ca:
        key: service-ca.crt
        configMapName: openshift-service-ca.crt
```

  pipelines:

>     - name: default-logstore
>       inputRefs:
>         - application
>         - infrastructure
>       outputRefs:
>         - default-lokiStack

Reviewer 2: Indentation as well:

```suggestion
  - name: default-logstore
    inputRefs:
    - application
    - infrastructure
    outputRefs:
    - default-lokiStack
```

----
<1> Specifies the ViaQ data model explicitly.
+
[NOTE]
====

> By default, the `dataModel` field is set to `""` (unset), which allows the Cluster Logging Operator (CLO) to determine the data model automatically. Currently, CLO defaults to the ViaQ model when the field is unset, but this might change in future releases. Specifying `dataModel: ViaQ` ensures the configuration remains compatible if the default changes.

Reviewer 5: ...but this will change in future releases.

====

.Verification
* Verify that logs are visible in the Log section of the *Observe* tab in the OpenShift web console.

=== Quick Start with OpenTelemetry

To configure OTLP ingestion and enable the OpenTelemetry data model, follow these steps:

:FeatureName: The OpenTelemetry Protocol (OTLP) output log forwarder
include::snippets/technology-preview.adoc[]

.Prerequisites
* Cluster administrator permissions

.Procedure

> . Install the `Red Hat OpenShift Logging`, `Loki Operator`, and `Cluster Observability Operator` operators from OperatorHub.

Reviewer 5: Shouldnt this say 'Red Hat Openshift Logging Operator'

. Create a `LokiStack` custom resource (CR) in the `openshift-logging` namespace:
+
[source,yaml]
----
apiVersion: loki.grafana.com/v1
kind: LokiStack
metadata:
  name: logging-loki
  namespace: openshift-logging
spec:
  managementState: Managed
  size: 1x.extra-small
  storage:
    schemas:
    - effectiveDate: '2024-10-01'
      version: v13
    secret:
      name: logging-loki-s3
      type: s3
    storageClassName: gp3-csi
  tenants:
    mode: openshift-logging
----
+
[NOTE]
====
Ensure that the `logging-loki-s3` secret is created beforehand. The contents of this secret vary depending on the object storage in use. For more information, see xref:../../../observability/logging/logging-6.0/log6x-upgrading-to-6.adoc#secrets-and-tls-configuration[Secrets and TLS Configuration].
====

. Create a service account for the collector:
+
[source,shell]
----
$ oc create sa collector -n openshift-logging
----

> . Bind the `ClusterRole` to the service account:

Reviewer 2: Is it better to instead describe what the role is for?

```suggestion
. Allow the collector's service account to write data to the LokiStack:
```

+
[source,shell]
----
$ oc adm policy add-cluster-role-to-user logging-collector-logs-writer -z collector
----
+
[NOTE]
====
The `ClusterRole` is created automatically during the Cluster Logging Operator installation and does not need to be created manually.
====

> . Add additional roles to the collector service account:

Reviewer 2: Is it better to instead describe what the role is for?

```suggestion
. Allow the collector's service account to collect logs:
```

> $ oc adm policy add-cluster-role-to-user collect-audit-logs -z collector

Author: Moving this step next to the step, "Bind the `ClusterRole` to the service account:"

  serviceAccount:
    name: collector
  outputs:

>   - name: default-lokiStack

Author: Revoked change here.

----

. Verify that logs are visible in the Log section of the Observe tab in the OpenShift web console.

> === Quick Start with OTEL

Reviewer 2: Do we need to duplicate the whole section or would it be more economic to just list the _differences_? I can see that some of the issues I have marked in the new content are just copied from the existing steps (they also need to be changed there) and I wonder if we can avoid this work in the future. Also it might be more digestable for users.

Author: I took the reference from Documentation Considerations mentioned in https://issues.redhat.com/browse/LOG-5637 in Jira description.

"The [default installation procedure](https://docs.openshift.com/container-platform/4.17/observability/logging/logging-6.0/log6x-about.html#quick-start) as described in 4.17+/6.0+ stays for as it, because we still want users to use the viaQ model as the currently GA format.

A new installation procedure copied from the [one in 4.17+/6.0+](https://docs.openshift.com/container-platform/4.17/observability/logging/logging-6.0/log6x-about.html#quick-start) needs to be established account for the dataModel field set to OTel as described here: https://github.com/openshift/cluster-logging-operator/blob/master/docs/features/logforwarding/outputs/redhat-managed-lokistack.adoc"

Please let me know if you have other thoughts.

To configure OTLP ingestion and enable the OTEL data model, follow these steps:

:FeatureName: The OpenTelemetry Protocol (OTLP) output log forwarder
include::snippets/technology-preview.adoc[]

.Prerequisites
* Cluster administrator permissions

.Procedure

> . Install the `OpenShift Logging` and `Loki` Operators from OperatorHub.

Reviewer 3: I believe this is the correct name
```suggestion
. Install the `Red Hat OpenShift Logging` and `Loki` Operators from OperatorHub.
```

Reviewer 2: These are the names from OperatorHub:

```suggestion
. Install the `Red Hat OpenShift Logging` and `Loki Operator` operators from OperatorHub.
```

. Create a `LokiStack` custom resource (CR) in the `openshift-logging` namespace:
+
[source,yaml]
----
apiVersion: loki.grafana.com/v1
kind: LokiStack
metadata:
  name: logging-loki
  namespace: openshift-logging
spec:
  managementState: Managed
  size: 1x.extra-small
  storage:
    schemas:

>     - effectiveDate: '2022-06-01'

Reviewer 2: Use a more recent date?

```suggestion
    - effectiveDate: 2024-10-01
```

      version: v13
    secret:
      name: logging-loki-s3

>       type: s3

Reviewer 2: This secret needs to be created as well and its contents depend on which object storage the customer is using.

Author: Added a Note for this.

    storageClassName: gp3-csi
  tenants:
    mode: openshift-logging
----

. Create a service account for the collector:
+
[source,shell]
----
$ oc create sa collector -n openshift-logging
----

> . Create a `ClusterRole` for the collector:

Reviewer 2: This ClusterRole is created during installation of the Cluster Logging Operator and does not need to be created manually.

Author: Maybe keep this step as Optional and add a Note for the same. I believe it will help the user to understand the flow better (else we can delete this step also). Please let me know if you have other thoughts. Thanks!

+
[source,yaml]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: logging-collector-logs-writer
rules:
- apiGroups:
  - loki.grafana.com
  resourceNames:
  - logs
  resources:
  - application
  - audit
  - infrastructure
  verbs:
  - create
----

. Bind the `ClusterRole` to the service account:
+
[source,shell]
----
$ oc adm policy add-cluster-role-to-user logging-collector-logs-writer -z collector
----

. Install the Cluster Observability Operator.

. Create a `UIPlugin` to enable the Log section in the Observe tab:
+
[source,yaml]
----
apiVersion: observability.openshift.io/v1alpha1
kind: UIPlugin
metadata:
  name: logging
spec:
  type: Logging
  logging:
    lokiStack:
      name: logging-loki
----

> . Add additional roles to the collector service account:

Reviewer 2: Do we need to explain what these roles are for?

Author: Added a Note.

+
[source,shell]
----
$ oc project openshift-logging
$ oc adm policy add-cluster-role-to-user collect-application-logs -z collector
$ oc adm policy add-cluster-role-to-user collect-audit-logs -z collector
$ oc adm policy add-cluster-role-to-user collect-infrastructure-logs -z collector
----

. Create a `ClusterLogForwarder` CR to configure log forwarding:
+
[source,yaml]
----
apiVersion: observability.openshift.io/v1
kind: ClusterLogForwarder
metadata:

>   name: my-logforwarder
>   namespace: my-app-namespace

Reviewer 2: We should put this into the `openshift-logging` namespace. Maybe it also makes sense to converge on a "default name" in the documentation. Previously we used `instance`, so we probably want to avoid that one. Maybe "clusterforwarder" or just "collector" are options?

```suggestion
  name: my-logforwarder
  namespace: openshift-logging
```

  annotations:
    observability.openshift.io/tech-preview-otlp-output: "enabled" # <1>
spec:
  outputs:

>     - type: lokistack # <2>

Reviewer 2: The type is `lokiStack`:

```suggestion
    - type: lokiStack # <2>
```

      name: loki-otlp
      lokiStack:
        target:

>           name: my-lokistack

Reviewer 2: This should match the name of the stack created above:

```suggestion
          name: logging-loki
```

          namespace: openshift-logging
        dataModel: Otel # <3>

>         tuning:
>           compression: none

Reviewer 2: I think we want to leave this out of the example and just make it use whatever default we decide to have.

```suggestion
```

        authentication:
          token:
            from: serviceAccount
      tls: # <4>
        ca:

>           key: service-ca.crt (7)
>           configMapName: openshift-service-ca.crt (7)

Reviewer 2: ```suggestion
          key: service-ca.crt
          configMapName: openshift-service-ca.crt
```

  pipelines:
   - name: my-pipeline
     inputRefs:
     - application
     - infrastructure
     outputRefs:
     - loki-otlp
  serviceAccount:
    name: logger-admin
----
<1> Use the annotation to enable the `Otel` output, which is a Technology Preview feature.

> <2> Define the `OTLP` output type as `lokistack`.

Reviewer 3: ```suggestion
<2> Define the output type as `lokistack`.
```

Reviewer 2: `lokiStack` also has a typo:

```suggestion
<2> Define the output type as `lokiStack`.
```

> <2> Define the `OTLP` output type as `lokistack`.

Author: Updated as, "Define the output type as `lokiStack`."

<3> DataModel is `Otel`.

> <4> `tls` spec identifies the keys and secret to the respective certificates that they represent.

Reviewer 2: I don't think this is valuable information as it does not explain anything. I don't think we want to generally explain the concept of secrets or TLS here.

Author: Removed callout # <4>.

+
[NOTE]
====

> You cannot use `lokistack` `labelKeys` when `dataModel` is `Otel`.

Reviewer 2: ```suggestion
You cannot use `lokiStack.labelKeys` when `dataModel` is `Otel`.
```

====

> . Verify that logs are visible in the Log section of the Observe tab in the OpenShift web console.

Reviewer 3: I believe a better way to check if OTLP was working correctly would be to check the "Distributor - Structured Metadata"  under the "OpenShift Logging / Lokistack / Writes" dashboard.

Author: Updated as:

* Verify that OTLP is functioning correctly by going to *Observe* > *OpenShift Logging* > *LokiStack* > *Writes* in the OpenShift web console, and checking *Distributor - Structured Metadata*.

These roles enable log collection across different log types: application, audit, and infrastructure logs.
====

. Create a `ClusterLogForwarder` CR to configure log forwarding:
+
[source,yaml]
----
apiVersion: observability.openshift.io/v1
kind: ClusterLogForwarder
metadata:
  name: collector
  namespace: openshift-logging
  annotations:
    observability.openshift.io/tech-preview-otlp-output: "enabled" # <1>
spec:
  outputs:
    - type: lokiStack # <2>
      name: loki-otlp
      lokiStack:
        target:
          name: logging-loki
          namespace: openshift-logging
        dataModel: Otel # <3>
        authentication:
          token:
            from: serviceAccount
      tls:
        ca:
          key: service-ca.crt
          configMapName: openshift-service-ca.crt
  pipelines:
   - name: my-pipeline
     inputRefs:
     - application
     - infrastructure
     outputRefs:
     - loki-otlp
  serviceAccount:
    name: logger-admin
----
<1> Use the annotation to enable the `Otel` output, which is a Technology Preview feature.
<2> Define the output type as `lokiStack`.
<3> DataModel is `Otel`.
+
[NOTE]
====
You cannot use `lokiStack.labelKeys` when `dataModel` is `Otel`.

> ====

Reviewer: Because we can't use the `lokistack.labelKeys` with `dataModel: Otel` anymore we should link and explain here that the `Configuring LokiStack for OTLP data ingestion` section is the new thing to accomplish the same task.

$ oc adm policy add-cluster-role-to-user collect-application-logs -z collector
$ oc adm policy add-cluster-role-to-user collect-audit-logs -z collector
$ oc adm policy add-cluster-role-to-user collect-infrastructure-logs -z collector
----
+
[NOTE]
====
These roles enable log collection across different log types: application, audit, and infrastructure logs.
====

. Create a `ClusterLogForwarder` CR to configure log forwarding:
+
[source,yaml]
----
apiVersion: observability.openshift.io/v1
kind: ClusterLogForwarder
metadata:
  name: collector
  namespace: openshift-logging
  annotations:
    observability.openshift.io/tech-preview-otlp-output: "enabled" # <1>
spec:
  outputs:
    - type: lokiStack # <2>
      name: loki-otlp
      lokiStack:
        target:
          name: logging-loki
          namespace: openshift-logging
        dataModel: Otel # <3>
        authentication:
          token:
            from: serviceAccount
      tls:
        ca:
          key: service-ca.crt
          configMapName: openshift-service-ca.crt
  pipelines:
   - name: my-pipeline
     inputRefs:
     - application
     - infrastructure
     outputRefs:
     - loki-otlp
  serviceAccount:

>     name: logger-admin

Reviewer: AFAICS, we speak about a collector service account named `collector` above, but we don't provide instructions to create one. Or is this linked somewhere else?

Author: I believe we have Step 3 for the same. Please let me know if you have other thoughts. Thanks!

Step 3:

Create a service account for the collector:

`oc create sa collector -n openshift-logging`

metadata:
  name: collector
  namespace: openshift-logging
  annotations:
    observability.openshift.io/tech-preview-otlp-output: "enabled" # <1>
spec:
  serviceAccount:
    name: logger-admin
  outputs:

>     - name: loki-otlp
>       type: lokiStack # <2>
>       lokiStack:
>         target:
>           name: logging-loki
>           namespace: openshift-logging
>         dataModel: Otel # <3>
>         authentication:
>           token:
>             from: serviceAccount
>       tls:
>         ca:
>           key: service-ca.crt
>           configMapName: openshift-service-ca.crt

Reviewer 2: Indentation again:

```suggestion
  - name: loki-otlp
    type: lokiStack # <2>
    lokiStack:
      target:
        name: logging-loki
        namespace: openshift-logging
      dataModel: Otel # <3>
      authentication:
        token:
          from: serviceAccount
    tls:
      ca:
        key: service-ca.crt
        configMapName: openshift-service-ca.crt
```
